Re-add dotnet docs. (#1210)
platformsh/platformsh-docs: src/languages/dotnet.md

# .NET

Platform.sh supports deploying .NET applications by allowing developers to define a build process and pass its variables to the .NET build environment.

## Supported versions

* 2.0
* 2.1

> * 2.2

Reviewer: Shouldn't this be using the new registry now? :smile:

Author: Absolutely, haven't gotten to pulling in the supportedHTML element yet on any of them.

Reviewer: Oh!  OK, I thought we'd done that part.

To specify a .NET container, use the `type` property in your `.platform.app.yaml`.

{% codesnippet "/registry/images/examples/full/dotnet.app.yaml", language="yaml" %}{% endcodesnippet %}

## Building the application

> .NET containers do not come with a pre-defined build process that you may have seen on other container types. Instead, the process is defined by you, using the runtime and output directory variables passed to the build environment.

Reviewer: I'm not sure this is relevant.  Only two container types, PHP and node, include "flavors" as we've been moving away them.  We don't have this comment on any other language AFAIK.

Author: Removing.

> For simple applications, we recommend using `dotnet publish --self-contained`, while specifying the target runtime and output folders via environment variables:

Reviewer: Avoid the use of first-person.  This can be reworded.  "For simple applications, `blah` is sufficient and allows you to specify..."  Or something like that.

hooks:
    build: |
        set -xe

>         dotnet publish --self-contained --runtime "$DOTNET_DEFAULT_RUNTIME" --output "$PLATFORM_OUTPUT_DIR" -p:UseRazorBuildServer=false -p:UseSharedCompilation=false

Reviewer: These aren't already set, are they?  Are users expected to set these?  If so, that should be documented very explicitly right here.  (I'm... not sure why you would want do do that via custom env vars frankly but that's another story.)

```

Typically .NET builds will start a collection of build servers, which are helpful for repeated builds. On Platform.sh, however, if this process is not disabled, the build process will not finish until the idle timeout is reached.

> Because of this it is recommended to include `-p` toggles that disable the Razor compiler for dynamic cshtml pages (`UseRazorBuildServer`) and the .NET msbuild compiler (`UseSharedCompilation`).

Reviewer: ```suggestion
As a result, it is recommended to include `-p` toggles that disable the Razor compiler for dynamic cshtml pages (`UseRazorBuildServer`) and the .NET msbuild compiler (`UseSharedCompilation`).
```

## Running the application

.NET applications should be started using the `web.commands.start` directive in `.platform.app.yaml`. This ensures that the command starts at the right moment and stops gracefully when a re-deployment needs to be executed. Also, should the program terminate for any reason, it will be automatically restarted. Note that the start command _must_ run in the foreground.

Incoming requests are passed to the application using either a TCP (default) or UNIX socket. The application must use the [appropriate environment variable](/configuration/app/web.html#socket-family) to determine the URI to listen on. In case of a TCP socket ([recommended](https://go.microsoft.com/fwlink/?linkid=874850)), the application must listen on `http://127.0.0.1`, using port defined by the `PORT` environment variable.

> There will be an Nginx server sitting in front of your application. We recommend serving static content via Nginx, as this allows easy control of headers (including cache headers), and also has marginal performance benefits. Note that HTTPS is also terminated at the Ngnix proxy, so the `app.UseHttpsRedirection();` line (which is in `Startup.cs` by default) should be removed. To force HTTPS-only, please refer to the [routes documentation](/configuration/routes/https.html#https).

Reviewer: The parenthetical here should not be parenthetical.  The sentence doesn't really make sense without it.  "the `blah` line in `Startup.cs` should be removed"

The following example configures your environment to serve the static content folders commonly found in ASP.NET MVC templates using Nginx, while routing other traffic to the .NET application.

```yaml
web:
  locations:
    "/":
      root: "wwwroot"
      allow: true
      passthru: true
      rules:
        # Serve these common asset types with customs cache headers.
        \.(jpe?g|png|gif|svgz?|css|js|map|ico|bmp|eot|woff2?|otf|ttf)$:
          allow: true
          expires: 300s

  commands:
    start: |

>       ./WebApplication1

Reviewer: No need to make this a multi-line block when it's a single line.

How does one control the output file name?

Author: Updated. From what I can tell it's taken from the name you give this file and how you define within it https://github.com/platformsh/template-builder/blob/aspnet/templates/aspnet-core/files/PlatformshAspNetCore.sln

Reviewer: Neither "Web" nor "Application" appear in that file. So, um, huh?

Author: Oh I'm just saying that that file would be called `WebApplication1.sln` with that name throughout this line 
```
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "PlatformshAspNetCore", "PlatformshAspNetCore.csproj", "{6150B0DB-45B5-4429-9410-2BA4E68874CC}"
```
so then that defines https://github.com/platformsh/template-builder/blob/446ccc0925747318576f765163a02b33190edf24/templates/aspnet-core/files/.platform.app.yaml#L28

Reviewer: Ah, ok.

```

If desired, you may also route all requests to the application unconditionally:

```yaml
web:
  locations:
    "/":
      allow: false
      passthru: true

  commands:
    start: |
    ./WebApplication1

> ```

Reviewer: Do we explain how to do this on any other language?

(Honestly we should revisit the language pages sometime soon.)

The following example configures an environment to serve the static content folders commonly found in [ASP.NET MVC](https://dotnet.microsoft.com/apps/aspnet/mvc) templates using Nginx, while routing other traffic to the .NET application.

```yaml
web:
  locations:
    "/":
      root: "wwwroot"
      allow: true
      passthru: true
      rules:
        # Serve these common asset types with customs cache headers.
        \.(jpe?g|png|gif|svgz?|css|js|map|ico|bmp|eot|woff2?|otf|ttf)$:
          allow: true
          expires: 300s

  commands:
    start: ./WebApplication1
```

You can also route all requests to the application unconditionally:

```yaml
web:
  locations:
    "/":
      allow: false
      passthru: true

  commands:
    start: ./WebApplication1
```

## Project templates

Platform.sh offers a project template for ASP.NET Core applications using the structure described above. It can be used as a starting point or reference for building your own application.

> [ASP.NET Core](https://github.com/platformsh/template-aspnet-core)

Reviewer: We should probably not merge this until that repo is populated.  Or remove this for now and re-add it when that template is resolved.

I'd favor the latter, I think.
